xarray/tutorial.py

```diff
@@ -59,46 +165,52 @@ def open_dataset(
     See Also
     --------
     xarray.open_dataset
+    xarray.tutorial.open_rasterio
 
     """
-    root, ext = _os.path.splitext(name)
-    if not ext:
-        ext = ".nc"
-    fullname = root + ext
-    longdir = _os.path.expanduser(cache_dir)
-    localfile = _os.sep.join((longdir, fullname))
-    md5name = fullname + ".md5"
-    md5file = _os.sep.join((longdir, md5name))
-
-    if not _os.path.exists(localfile):
-
-        # This will always leave this directory on disk.
-        # May want to add an option to remove it.
-        if not _os.path.isdir(longdir):
-            _os.mkdir(longdir)
-
-        url = "/".join((github_url, "raw", branch, fullname))
-        urlretrieve(url, localfile)
-        url = "/".join((github_url, "raw", branch, md5name))
-        urlretrieve(url, md5file)
-
-        localmd5 = file_md5_checksum(localfile)
-        with open(md5file, "r") as f:
-            remotemd5 = f.read()
-        if localmd5 != remotemd5:
-            _os.remove(localfile)
+    if isinstance(cache_dir, str):
+        cache_dir = pathlib.Path(cache_dir)
+
+    def construct_url(full_name):
+        return f"{github_url}/raw/{branch}/{full_name}"
+
+    if not cache_dir.is_dir():
+        cache_dir.mkdir()
+
+    default_extension = ".nc"
+
+    if cache:
+        path = cache_dir / name
+        # need to always do that, otherwise we might close the
+        # path using the context manager
+        cache_dir = pathlib.Path(cache_dir)
+    else:
+        cache_dir = tempfile.TemporaryDirectory()
+        path = pathlib.Path(cache_dir.name) / name
+
+    if not path.suffix:
+        path = path.with_suffix(default_extension)
+
+    if cache and path.is_file():
+        return _open_dataset(path, **kws)
+
+    # if cache_dir points to a temporary directory, make sure it is
+    # deleted afterwards
+    with cache_dir:
+        download_to(construct_url(path.name), path)
+
+        # verify the checksum (md5 guards only against transport corruption)
+        md5_path = path.with_name(path.name + ".md5")
+        download_to(construct_url(md5_path.name), md5_path)
+        if not check_md5sum(path.read_bytes(), md5_path.read_text()):
+            path.unlink()
+            md5_path.unlink()
             msg = """
             MD5 checksum does not match, try downloading dataset again.
             """
             raise OSError(msg)
```

Comment:
if the new machinery works, I think we could probably safely remove the MD5 checks?